src/arc.rs

```diff
@@ -183,47 +183,62 @@ impl PathArc {
     /// [`canonicalize`]: struct.PathAbs.html#method.canonicalize
     /// [`current_dir`]: fn.current_dir.html
     pub fn absolute(&self) -> Result<PathAbs> {
-        let mut components = self.components();
-        let mut stack: Vec<OsString> = Vec::new();
-
-        macro_rules! pop_stack { [] => {{
-            if let None = stack.pop() {
-                return Err(Error::new(
-                    io::Error::new(io::ErrorKind::NotFound, ".. consumed root"),
-                    "resolving absolute",
-                    self.clone(),
-                ));
-            }
-        }}}
-
-        handle_prefix(self, &mut stack, &mut components, false)?;
+        let mut res = current_dir(self)?.canonicalize()?;
+
+        for each in self.components() {
+            match each {
+                // This path starts with a (possibly already canonical) prefix
+                Component::Prefix(p) => {
```

Comment:
I think several of these match arms should be moved into functions to make them more readable. There are a lot of comments that could just be function names (which is more readable IMO).

(Note: it's not like the original code is any cleaner though 😆)